pkg/dns/dns.go

```diff
@@ -458,13 +509,41 @@ func (kd *KubeDNS) Records(name string, exact bool) (retval []skymsg.Service, er
 
 	// If the name query is not an exact query and does not match any records in the local store,
 	// attempt to send a federation redirect (CNAME) response.
-	if !exact {
-		return kd.federationRecords(path)
+	if !exact && isFederationQuery {
+		glog.Infof("federation service query: Did not find a local service. Trying federation redirect (CNAME) response")
+		return kd.federationRecords(reverseArray(federationSegments))
 	}
 
 	return nil, etcd.Error{Code: etcd.ErrorCodeKeyNotFound}
 }
 
+// Returns true if the service corresponding to the given message has endpoints.
+// Always returns true for headless services.
```

Comment:
I don't understand why you want to return true for headless services, even if they have no endpoints?  Please explain.